plutus-use-cases/src/Language/Plutus/Coordination/Plutus.hs

```diff
@@ -0,0 +1,231 @@
+-- | A model of the types involved in transactions, and of the wallet API.
+module Language.Plutus.Coordination.Plutus (-- * Transactions and related types
+                Address
+              , PubKey(..)
+              , KeyPair
+              , pubKey
+              , Value
+              , Tx(..)
+              , TxIn(..)
+              , TxOut(..)
+              , mkAddress
+              , TxOutRef(..)
+              , standardTxFee
+              , txOutValue
+              , txOutDataScript
+              , txOutValidatorScriptHash
+              -- * API operations
+              , TxM
+              , Hash
+              , hash
+              , Redeemer
+              , Validator
+              , DataScript
+              , PlutusTx(..)
+              , BlockHeight
+              , PendingTx(..)
+              , submitTransaction
+              , assert
+              , lookupMyKeyPair
+              , lookupMyPubKey
+              , createPayment
+              , txInSign
+              , Range(..)
+              , EventTrigger(..)
+              , Signed(..)
+              , OracleValue(..)
+              ) where
+
+import           Control.Applicative              (Alternative (..))
+import           Control.Monad.State              (State)
+import           Control.Monad.Trans.Maybe        (MaybeT (..))
+import           Language.Plutus.CoreToPLC.Plugin (PlcCode)
+
+newtype Signed a = Signed (PubKey, a)
+
+-- `(pk, (bh, i, a)) :: OracleValue a` is the value of an "observable" `i`
+-- (sensor, etc.) observed at `bh` signed by an oracle `pk`.
+--
+newtype OracleValue a = OracleValue (Signed (BlockHeight, Int, a))
+
+-- | Cardano address
+--
+newtype Address = Address Int
+    deriving (Eq, Ord, Show, Read)
+
+-- | Ada value
+--
+type Value = Int
+
+newtype Hash = Hash Int
+    deriving (Eq, Ord, Show, Read)
+
+hash :: a -> Hash
+hash _ = Hash 10
+
+-- | Public key
+--
+data PubKey = PubKey
+
+-- | Public key pair (no lift instance, because we never ought to put it into a
+--   transaction)
+--
+data KeyPair
+
+data TxState
+
+-- | Transaction monad for coordination layer computations; provides access to
+-- the blockchain
+--
+type TxM a = MaybeT (State TxState) a
+
+-- | Submit the given transaction to the blockchain
+submitTransaction :: Tx -> TxM [TxOutRef]
+submitTransaction = const empty
+
+-- | Verify that a condition is true.
+assert :: Bool -> TxM ()
+assert = const empty
+
+-- | Get the users's public key. Part of the wallet interface
+lookupMyPubKey :: TxM PubKey
+lookupMyPubKey = pubKey <$> lookupMyKeyPair
+
+-- | Extract the public key from a key pair.
+pubKey :: KeyPair -> PubKey
+pubKey = const PubKey
+
+-- | Part of the wallet interface
+--   TODO: Should the Plutus client even be able to know the private key?
+lookupMyKeyPair :: TxM KeyPair
+lookupMyKeyPair = empty
+
+-- | Create an input that spends the given value (part of the wallet interface)
+--
+createPayment :: Value -> TxM TxIn
```

Comment:
I don't entirely see how this can work - shouldn't this need to be a `TxOut`?

Reply:
This and `txInSign` seem to be for the "non-scripting" version of transactions. The paper doesn't explicitly say how to make them coexist - are we expecting to put in "trivial" scripts to make this sort of thing work?

Reply:
> are we expecting to put in "trivial" scripts to make this sort of thing work?

That would make the model easier but it means that the size of all non-scripted transactions increases.

Reply:
> shouldn't this need to be a TxOut?

Yeah but then in order to spend it in a subsequent transaction we would need a `TxIn` referencing the output. So the signature is just for convenience.

Reply:
Hm I see this makes sense because it's theoretically part of the wallet API. It would be nice if we required such things to take an explicit `Wallet` argument or run in `WalletApi m`, so it's clear that something else is happening. Then e.g. it would be obvious that the code run by the agents in the examples needs to run in a wallet.